1. What is Python?
Definition

Python is an interpreted, high‑level, general‑purpose programming language. 
datagy
+2
dshub.gitbook.io
+2

Interpreted means you write code and it runs (via an interpreter) without a separate compile step (in many cases).

High‑level means that it abstracts away many of the details of how the computer works (memory management, etc) so you can focus on logic.

General‑purpose means you can use it for many applications: web development, scripting, automation, data analysis, AI, etc.

It is dynamically typed (you don’t need to declare variable types ahead) and has built‑in memory management (garbage collection). 
datagy

Python’s syntax emphasizes readability and simplicity, which makes it good for beginners.

-----------------------
History/Why use Python
-----------------------
Created by Guido van Rossum in the late 1980s, first released in 1991.

It became popular because of its simplicity, large standard library, ecosystem of external libraries, and versatility.

For beginners: less boilerplate and more immediate feedback — you can write stuff quickly, see results.

What can you do with Python?

Automate tasks (e.g., file operations, web scraping)

Build websites (with frameworks like Django, Flask)

Data analysis and visualization (with pandas, NumPy, matplotlib)

Scientific computing, machine learning

Scripting, command‑line tools, and more

-----------------
Summary for you
-----------------
If you’re just starting, think of Python as a friendly programming language you can quickly jump into: you write code, see results, explore, experiment — fewer hurdles than many older languages.

-----------------------------------------------
2. How to install Python & set up environment
-----------------------------------------------
Here’s a beginner‑friendly step‑by‑step guide. (Also applies for your system in Guntur / India; instructions are generic.)

Step‑1: Download

Go to the official website: python.org

Under “Downloads” select your operating system (Windows, macOS, Linux).

Step‑2: Install

Windows: run the installer .exe. Make sure you check “Add Python to PATH” if offered.

macOS: use the installer .pkg or use Homebrew (brew install python3).

Linux: often Python comes pre‑installed. You can install latest via sudo apt install python3 (Ubuntu) etc.

------------------------------
Step‑3: Verify installation
------------------------------
Open a terminal or command prompt and type python --version or python3 --version.

It should show something like Python 3.10.x (or whatever version you installed).

Then type python (or python3) and you’ll get a REPL (interactive shell) prompt >>>.

Try print("Hello, world!").

---------------------------------
Step‑4: Choose an editor or IDE
---------------------------------
For beginners, it’s good to use either:

A code editor like Visual Studio Code with Python extension

Or a notebook environment like Jupyter Notebook / JupyterLab

Or an IDE like PyCharm

For interactive experimentation (especially if you later work with the GitHub notebook), Jupyter is very convenient.

-------------------
Step‑5: (Optional) Use virtual environment
-------------------
To keep your Python projects isolated (so one project’s libraries don’t conflict with another), you can use venv in Python 3:
```bash
python3 -m venv myenv
source myenv/bin/activate  # (on Windows: myenv\Scripts\activate)

```

----------------
Step‑6: Install Jupyter Notebook
---------------
If you want to open the notebook from GitHub, you might install:
```bash
pip install notebook
jupyter notebook
```

That will open a browser where you can open .ipynb files and run code cells interactively.

---------------------
Code Explanation:--
---------------------

Printing / output: Use print(...) to show text or data on the screen.

Comments: Lines beginning with # are comments (ignored by interpreter) → used for explanation.

In [1]:
print("hello world")

hello world


Code Explanation:
Basic arithmetic / expressions: e.g., 2 + 3, 10 * (5‑2), etc.

You’ll see that Python evaluates expressions and you get results immediately.

In [2]:
print(3+7)

10


In [3]:
4+9

13

Code Explanation:
---
variable = 80 - 2:
This assigns the result of the expression 80 - 2 (which is 78) to a variable named variable.

variable is the name.

= is the assignment operator.

80 - 2 is an arithmetic expression (subtraction).

variable:
This line prints the current value stored in variable, which is 78. In Jupyter Notebook, writing the variable name alone on a line displays its value.

In [4]:
variable = 80-2
variable

78

Code Explanation:
---
type() is a built-in Python function that tells you what kind of data (or data type) a value or variable holds.

In this case, variable contains the result of 80 - 2, which is 78.

Since 78 is a whole number, type(variable) will return <class 'int'>.

🧠 Takeaway: Use type() to understand the type of value you're working with—important for debugging and learning.

In [5]:
type(variable)

int

Code Explanation:
---
Multiple Assignment: This line assigns three values to three variables in a single line:

a = 1

b = 2

c = 3

This is called tuple unpacking (even without parentheses). Python assigns each value on the right to the corresponding variable on the left.

The second line (a, b, c) returns a tuple: (1, 2, 3) when run in a Jupyter Notebook.

In [6]:
a,b,c = 1,2,3
a,b,c

(1, 2, 3)

Code Explanation:
---
import keyword:
This imports Python's built-in keyword module, which helps you work with reserved words in the language.

keyword.kwlist:
This returns a list of all keywords in Python — words like if, else, for, while, def, class, etc., that cannot be used as variable names.

In [7]:
import keyword
print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


Code Explanation:
---
keyword.kwlist returns a list of all Python keywords.

len() is a built-in function that returns the number of items in a list.

So, len(keyword.kwlist) tells you how many keywords exist in your current version of Python.

🧠 Example output (as of Python 3.10+): It will return something like 36, but this number may change with Python versions.

In [8]:
len(keyword.kwlist)

35

Code Explanation:
---
This line creates a variable named var_ and assigns the value 6 to it.

Variable names can include letters, numbers, and underscores (_), but cannot start with a number.

Here, var_ is a valid variable name because it starts with a letter and ends with an underscore.

The value 6 is an integer.

✅ Tip: Use meaningful variable names to make your code easier to understand.

In [9]:
var_=6

Code Explanation:
---
This creates a variable named _var and assigns the value 6 to it.

Variable names in Python can start with an underscore (_), which is perfectly valid.

Starting a variable with _ is sometimes used to indicate it's intended for internal use or "private" (by convention, not enforced by Python).

The value 6 is an integer (int).

✅ So _var is a valid variable name, just like var_ or var1.

In [10]:
_var=6

Code Explanation:
---
z = 8
Assigns the integer value 8 to the variable z.

Z = z
Assigns the value stored in z (which is 8) to a new variable Z.
Note: Variable names are case-sensitive in Python, so z and Z are two different variables.

print(Z)
Displays the value of Z in the output, which is 8.

In [11]:
z=8
Z=z
print(Z)

8


Code Explanation:
---
key = "string content"
This assigns the string "string content" to the variable named key.

A string is a sequence of characters enclosed in quotes (single '...' or double "..." quotes).

key (on its own line) in a Jupyter Notebook outputs the current value of the variable, which is the string "string content".

In [12]:
key="string content"
key

'string content'

Code Explanation:
---
The type() function tells you the data type of the value stored in a variable.

Since key holds "string content", which is text enclosed in quotes, type(key) will return <class 'str'>.

<class 'str'> means the variable is a string.

In [13]:
type(key)

str

Code Explanation:
---
This is a simple division operation in Python.

The / operator divides the number on the left (20) by the number on the right (4).

The result is 5.0 — note that Python returns a floating-point number (decimal) even if the division is exact.

In [14]:
20/4

5.0

Code Explanation:
---
This divides 100 by 6 using the / operator.

Since the division doesn’t result in a whole number, Python returns a floating-point number (a decimal).

The output will be approximately 16.666666666666668.

In [15]:
100/6

16.666666666666668

In [16]:
100//6

16

Code Explanation:
---
This divides 20 by 2 using the / operator.

The result is 10.0 — a floating-point number (decimal), even though the division is exact.

Python’s division operator / always returns a float (decimal), never an integer.

In [17]:
20/2

10.0

Code Explanation:
---
// is the floor division operator in Python.

It divides the number on the left by the number on the right and then rounds down to the nearest whole number (integer).

In this case, 20 // 2 divides 20 by 2, which equals exactly 10, so the result is 10 (an integer, not a float).

In [18]:
20//2

# floor division

10

Code Explanation:
---
% is the modulus operator in Python.

It returns the remainder after dividing the left number by the right number.

In this case, 25 % 5 means: divide 25 by 5, then return the remainder.

Since 25 is exactly divisible by 5, the remainder is 0.

In [19]:
25%5

# modulus

0

Code Explanation:
---
* is the multiplication operator in Python.

This multiplies the number on the left (2) by the number on the right (3).

The result is 6.

In [20]:
2 * 3


6

Code Explanation:
---
** is the exponentiation operator in Python.

It raises the number on the left (2) to the power of the number on the right (3).

So, 2 ** 3 means 

=2×2×2=8.

In [21]:
2 ** 3

#power of 2

8

Code Explanation:
---
a = 5: Assigns the value 5 to the variable a.

a += 5: This is a shortcut for a = a + 5. It adds 5 to the current value of a and updates a.

print(a): Displays the current value of a.

In [22]:
a =5 
a+= 5
print(a)

10


a = 5: Assigns the value 5 to the variable a.

a -= 2: This is shorthand for a = a - 2. It subtracts 2 from the current value of a and updates a.

print(a): Prints the new value of a.

In [23]:
a = 5
a-=2
print(a)

3


a = 4: Assigns the value 4 to the variable a.

a == 4: This is a comparison operator that checks if a is equal to 4.

The == operator does NOT assign a value; it compares two values.

The result of a == 4 is a Boolean value:

True if a equals 4

False otherwise

In [24]:
a=4

#comparision operator
a==4  

True

a = 5: Assigns the value 5 to variable a.

a += 5: Adds 5 to the current value of a.
So, now a is 10 (5 + 5).

a == 5: Checks if the current value of a is equal to 5.

Since a is now 10, the expression a == 5 evaluates to False.

In [25]:
a =5

a+=5

#comparision operator
a==5

False

a = 5: Assigns the value 5 to variable a.

a > 3: This is a comparison operator that checks if a is greater than 3.

Since 5 is greater than 3, this expression evaluates to True.

In [26]:
a =5

a > 3

True

a = 5: Assigns the value 5 to the variable a.

a > 20: Checks if a is greater than 20.

Since 5 is not greater than 20, this expression evaluates to False.

In [27]:
a = 5

a > 20

False

a = 5: Assigns the value 5 to the variable a.

a < 20: Checks if a is less than 20.

Since 5 is less than 20, this expression evaluates to True.

In [28]:
a = 5

a < 20

True

a = 5: Assigns the value 5 to the variable a.

a <= 3: Checks if a is less than or equal to 3.

Since 5 is greater than 3, this expression evaluates to False.

In [29]:
a = 5

a <= 3

False

a = 5: Assigns the value 5 to the variable a.

a >= 1: Checks if a is greater than or equal to 1.

Since 5 is greater than 1, this expression evaluates to True.

In [30]:
a = 5

a >= 1

True

a = 5 and b = 5: Assigns the value 5 to variables a and b.

a > 3: Checks if a is greater than 3 → True because 5 > 3.

b > 3: Checks if b is greater than 3 → True.

a > 3 and b > 3: Uses the logical AND operator to check if both conditions are true.

Since both a > 3 and b > 3 are True, the whole expression is True.

Quick note on and:

and returns True only if both conditions are true.

If either one is false, it returns False.


| a > 3 | b > 3 | a > 3 and b > 3 |
| ----- | ----- | --------------- |
| True  | True  | True            |
| True  | False | False           |
| False | True  | False           |
| False | False | False           |


In [31]:
a = 5
b = 5

# a>3  
# true

b>3
# b > 3
# True

a>3 and b>3

True

a = 5 and b = 5: Assigns the value 5 to both variables.

a > 3 and b < 3: This uses the logical AND operator.

It checks two conditions:

a > 3 → Is 5 greater than 3? True

b < 3 → Is 5 less than 3? False

Since AND requires both conditions to be True, and one is False, the whole expression evaluates to False.

Summary of and operator:

Returns True only if both conditions are True.

Returns False if any one condition is False.

In [32]:
a = 5
b = 5

a>3 and b<3

# and ....operator

False

a = 5 and b = 0: Assigns values to variables a and b.

a > 1 or b > 1: Uses the logical OR operator to check if at least one condition is true.

a > 1 → Is 5 greater than 1? True

b > 1 → Is 0 greater than 1? False

Since OR requires only one condition to be true, the entire expression evaluates to True.


Summary of or operator:

Returns True if at least one condition is True.

Returns False only if both conditions are False.


| a > 1 | b > 1 | a > 1 or b > 1 |
| ----- | ----- | -------------- |
| True  | True  | True           |
| True  | False | True           |
| False | True  | True           |
| False | False | False          |


In [33]:
a = 5
b = 0

a>1 or b>1

# or operator

True

Bitwise operator

a = 5 and b = 3: Assign values 5 and 3 to variables a and b.

a & b: This is the bitwise AND operator.

It performs a bit-by-bit comparison of the binary forms of a and b and returns a new number.

How it works:

Binary of 5 is 0101

Binary of 3 is 0011

  0101  (5)
& 0011  (3)
= 0001  (1)


In [34]:
a = 5
b = 3
a  & b

1

a = 5 and b = 3: Assign values 5 and 3 to variables a and b.

a | b: This is the bitwise OR operator.

It compares the binary representation of a and b bit by bit and returns a new number.

The result has bits set to 1 if either of the bits in a or b is 1.

How it works:

Binary of 5 is 0101

Binary of 3 is 0011

Perform bitwise OR (|):
  0101  (5)
| 0011  (3)
= 0111  (7)


In [35]:
a = 5
b = 3
a | b

7

a = "chandra": Assigns the string "chandra" to variable a.

"c" in a: Checks if the character "c" is present anywhere in the string a.

Since "c" is the first letter of "chandra", the expression returns True.

print() displays the result, so it will print True.

Summary:

The keyword in can be used to check if a substring or character exists inside a string (or other collections like lists).

Returns True if found, otherwise False.




In [36]:
a = "chandra"
print("c" in a)

True


a = "chandra": Assigns the string "chandra" to the variable a.

"zyx" in a: Checks if the substring "zyx" is present anywhere inside the string a.

Since "zyx" is not found in "chandra", the expression evaluates to False.

The print() statement outputs False.

Summary:

The in operator checks if one string is a substring of another.

If the substring is not found, it returns False.

In [37]:
a = "chandra"

print("zyx" in a)

False


a = [1, 2, 3, 4]: Creates a list with four numbers.

10 in a: Checks if the number 10 is present in the list a.

Since 10 is not in the list [1, 2, 3, 4], the expression returns False.

print() outputs False.

Summary:

The in operator works with lists, strings, and other collections to check for the presence of an element.

Returns True if the element is found, otherwise False.

In [38]:
a = [1,2,3,4]

print(10 in a)

False


a = True: Assigns the Boolean value True to variable a.

not a: Uses the logical NOT operator.

not reverses the Boolean value:

If a is True, then not a is False.

If a is False, then not a is True.

Summary:

not flips the truth value.

It's useful when you want to check the opposite of a condition.

In [39]:
a = True

not a

#not operator

False

'Math' module

import math: This imports Python’s built-in math module, which contains many useful mathematical functions and constants.

math.pi: Accesses the constant π (pi) from the math module.

math.pi gives you the value of π (approximately 3.141592653589793), which is useful for calculations involving circles, angles, etc.

Summary:

The math module provides handy math tools like pi, sqrt(), sin(), and more.

To use these tools, you first import the module with import math.

In [40]:
import math

math.pi

3.141592653589793

import math: Imports the math module to use math constants and functions.

radius = 5: Assigns the value 5 to the variable radius, representing the radius of a circle.

area = math.pi * radius ** 2: Calculates the area of a circle using the formula 
𝜋
𝑟
2
πr
2
:

math.pi gives the value of π (approximately 3.14159).

radius ** 2 means radius squared (radius multiplied by itself).

print(area): Displays the calculated area.

What happens step-by-step:

Square the radius: 
5 x 5 = 25


Multiply by π: 
25
×
3.14159
≈
78.54
25×3.14159≈78.54.

The output will be approximately 78.53981633974483.

Summary:

This program calculates and prints the area of a circle with radius 5.

You can change the radius value to calculate the area for other circles.

In [ ]:
import math

radius = 5
area = math.pi * radius ** 2  # Area of a circle
print(area)

import math: Imports the built-in math module.

math.nan: Represents the special floating-point value NaN, which stands for "Not a Number".

NaN is used to represent undefined or unrepresentable numeric results, such as:

Dividing 0 by 0

Taking the square root of a negative number (in some contexts)

Invalid operations in floating-point calculations

Summary:

math.nan is a special value to indicate an invalid or undefined number in calculations.

It is useful in scientific computing and data processing to signal missing or erroneous numerical data.

In [41]:
import math

math.nan

#not a number

nan

import math: Imports Python’s math module.

math.floor(200.34567): Returns the floor value of the number 200.34567.

Floor means rounding down to the nearest whole integer.

So, math.floor(200.34567) returns 200.

Summary:

The floor function always rounds a decimal number down to the closest integer.

Even if the decimal part is .999, it rounds down.

In [42]:
import math

math.floor(200.34567)

# floor value

200

import math: Imports Python’s math module.

math.ceil(300.999): Returns the ceiling value of the number 300.999.

Ceiling means rounding up to the nearest whole integer.

So, math.ceil(300.999) returns 301.

Summary:

The ceil() function always rounds a decimal number up to the closest integer.

Even if the decimal part is very small (e.g., 300.0001), it rounds up.

In [43]:
import math

math.ceil(300.999)

# ceil value

301

import math: Imports Python’s math module.

math.log(100): Calculates the natural logarithm (logarithm base e) of 100.

The natural logarithm is the power to which e (approximately 2.71828) must be raised to get the number.

So, math.log(100) answers: "e to the power of what equals 100?"

Summary:

math.log(x) returns the natural logarithm of x.

If you want the logarithm with base 10, use math.log10(x) instead.

In [44]:
import math

math.log(100)

4.605170185988092

import math: Imports Python’s math module.

math.sqrt(16): Calculates the square root of 16.

The square root of a number is a value that, when multiplied by itself, gives the original number.

So, the square root of 16 is 4 because 
4
×
4
=
16
4×4=16.

Summary:

Use math.sqrt(x) to find the square root of x.

It returns a floating-point number (decimal).

In [45]:
import math

math.sqrt(16)

4.0

import math: Imports Python’s math module.

math.pow(2, 3): Calculates 2 raised to the power of 3, which means 


This computes the cube of 2, i.e., 
2×2
×
2
=
8
........
2×2×2=8.

The result is returned as a floating-point number (8.0).

Summary:

math.pow(base, exponent) raises base to the power of exponent.

For integer powers, you can also use the ** operator, e.g., 2 ** 3.

In [46]:
import math

math.pow(2,3)

# cube of 2

8.0

a = [1, 2, 3, 4]: Creates a list containing the numbers 1, 2, 3, and 4.

10000 not in a: Checks if 10000 is NOT present in the list a.

Since 10000 is not in the list, this expression evaluates to True.

print() displays the result, so it will print True.

Summary:

The not in operator checks if an element is absent from a list (or other collections).

Returns True if the element is not found; otherwise False.

In [47]:
a = [1,2,3,4]

print(10000 not in a)

True


a = 5: Assigns the value 5 to variable a.

if a == 5: This is an if statement that checks if a is equal to 5.

The condition a == 5 evaluates to True because a is indeed 5.

If the condition is True, the indented block of code below the if statement runs.

print(a) will execute and print the value of a, which is 5.

Summary:

The if statement is used to run code based on a condition.

If the condition is True, the indented block of code under if will run.

In [48]:
a = 5

if a == 5:
    print(a)

5


a = 5: Assigns the value 5 to the variable a.

if a > 10000: Checks if a is greater than 10,000.

Since 5 is not greater than 10,000, the condition is False.

Because the condition is False, the code inside the else block will run.

print("false") will execute, so it will print false.

Summary:

The if statement checks a condition and runs the code inside it if the condition is True.

The else block runs if the condition is False.

In [49]:
a = 5

if a > 10000:
    print("true")

else:
    print("false")

false


a = 5: Assigns the value 5 to variable a.

if a > 0: Checks if a is greater than 0.

Since a = 5, which is greater than 0, the condition is True, so the code inside the if block runs.

Inside the if block:

a = a + 3: Adds 3 to a, changing the value of a from 5 to 8.

print(a): Prints the updated value of a, which is now 8.

The else block is not executed because the condition a > 0 was True.

Summary:

The if statement is used to check if a condition is True and run the code inside it.

If the condition is False, the code inside the else block will execute instead.

In this case, the condition was True, so the value of a was updated and printed as 8.

In [50]:
a = 5

if a > 0:
    a = a + 3
    print(a)

else:
    print(a)

8


a = 5: Assigns the value 5 to the variable a.

if a == 5: Checks if a is equal to 5.

This condition is True, so the code inside the if block runs.

a = a + 5: Increases the value of a by 5. Now a becomes 10.

print(a): Outputs the updated value of a, which is 10.

In [51]:
a = 5

if a == 5:
    a = a + 5
    print(a)

10


a = 5: The variable a is assigned the value 5.

if a == 1000: Checks if a is equal to 1000.

This condition is False because a is 5, not 1000.

Therefore, the code inside the if block is skipped.

print("outside of if-block"): This line is outside the if block and runs no matter what, so it prints:
....................
outside of if-block
...................

In [52]:
a = 5

if a == 1000:
    print("if statement is true")

print("outside of if-block")

outside of if-block


a = 5: Variable a is assigned the value 5.

if a == 3: This checks if a is equal to 3. It is False, so this block is skipped.

elif a == 5: This checks if a equals 5. It is True, so this block runs:

"elif statements get executed" is printed.

Then a is printed (which is 5).

The rest of the elif blocks are skipped after the first one is True.

In [53]:
a = 5

if a == 3:
    print(a+3)

elif a ==5:
    print("elif statements get executed")
    print(a)

elif a > 20:
    print("nested elif")

elif statements get executed
5


a = 5: Variable a is assigned the value 5.

if a > 200: Checks if a is greater than 200. This is False.

elif a < 0: Checks if a is less than 0. This is also False.

Since both conditions are false, the else block runs.
--------
Explanation of pass:
--------
pass is a placeholder statement in Python. It does nothing and is used when you need a block syntactically but don’t want it to do anything (yet).
-----------
else block statements get executed
5


In [54]:
a = 5

if a > 200:
    pass

elif a < 0:
    pass

else:
    print("else block statements get executed")
    print(a)

else block statements get executed
5


1. print(str.strip())

strip() removes leading and trailing spaces (spaces at the start and end).

✅ Output: "a chandra saibfg b c"
-------------

2. print(str.lstrip())

lstrip() removes only leading spaces (on the left).

✅ Output: "a chandra saibfg b c "
------------


3. print(str.replace("sai", "sekhar"))

Replaces every occurrence of "sai" with "sekhar".

"chandra saibfg" becomes "chandra sekharbfg".

✅ Output: " a chandra sekharbfg b c "
-----------

4. print(str.strip(' '))

Same as strip(). Here, ' ' (a space) is passed explicitly — it removes spaces from both ends.

✅ Output: "a chandra saibfg b c"
----------------

5. print(str.split('a'))

Splits the string at every 'a', removing the 'a' and returning a list of parts.

✅ Output: A list like: [' ', ' ch', 'ndr', ' s', 'ibfg b c ']
----------------

6. print(str.split(' '))

Splits the string at every space. Since there are many consecutive spaces, it creates many empty strings.

✅ Output: A list with words and many '' (empty strings), like:

['', '', '', 'a', '', '', 'chandra', 'saibfg', '', '', 'b', '', '', '', 'c', '', '', '', '']
------------------

7. print(f"max value of str: {max(str)}")

Finds the max character based on Unicode (alphabetical and symbol order).

Since space (' ') has a low value, and 's' or 'r' have higher ones, it prints the highest character.

✅ Output: "max value of str: s"
----------------------

8. print(str)

Simply prints the original string.

✅ Output: ' a chandra saibfg b c '
-----------------------

9. print("minimum value", min(str.split()))

First: str.split() splits the string by default space (ignoring extra spaces).

Then: min() finds the smallest word alphabetically.

The words in the list will be: ['a', 'chandra', 'saibfg', 'b', 'c']

'a' is the smallest alphabetically.

✅ Output: "minimum value a"

In [55]:
str="   a   chandra saibfg   b    c    "
print(str.strip())
print(str.lstrip())
print(str.replace("sai", "sekhar"))
print(str.strip(' '))
print(str.split('a'))
print(str.split(' '))
print(f"max value of str: {max(str)}")
print(str)
print("minimum value",min(str.split()))

a   chandra saibfg   b    c
a   chandra saibfg   b    c    
   a   chandra sekharbfg   b    c    
a   chandra saibfg   b    c
['   ', '   ch', 'ndr', ' s', 'ibfg   b    c    ']
['', '', '', 'a', '', '', 'chandra', 'saibfg', '', '', 'b', '', '', '', 'c', '', '', '', '']
max value of str: s
   a   chandra saibfg   b    c    
minimum value a


word = ['chandra', ':', 'sai']: This is a list of strings.

"_".join(word): This joins all elements in the list using the underscore (_) as a separator.

It creates one single string by placing _ between each item.

In [56]:
word = ['chandra', ':', 'sai']
"_".join(word)

'chandra_:_sai'

fruits is a list of strings: ["apple", "banana", "orange"].

if "chandra" in fruits:

Checks if "chandra" exists in the fruits list.

It does not, so this condition is False.

elif "cherry" in fruits:

Checks if "cherry" exists in the list.

It also does not, so this is False.

else:

Since none of the conditions were True, the else block is executed.

✅ Output:
else block got executed

In [57]:
fruits = ["apple", "banana", "orange"]

if "chandra" in fruits:
    print("chandra is present in the list")

elif "cherry" in fruits:
    print("cherry is present in list")

else:
    print("else block got executed")

else block got executed


List

a = [1, 2, 3, 4, 5, 6]: A list containing six numbers is assigned to variable a.

len(a): The built-in len() function returns the number of elements in the list.

✅ Output:

6

In [58]:
a = [1,2,3,4,5,6]
len(a)

6

a = [1, 2, 3, 4, 5, 6]: A list with 6 numbers.

a[0:2]: This is called list slicing.

It extracts elements from index 0 up to but not including index 2.

So it grabs elements at index 0 and 1, which are 1 and 2.

In [59]:
a =[1,2,3,4,5,6]

a[0:2]

[1, 2]

a = [1, 2, 3, 4, 5, 6]: A list with 6 numbers.

a[:4] is a slice of the list.

When the start index is omitted, it means start from the beginning (index 0).

It selects elements from the beginning up to but not including index 4.

So it includes elements at indices 0, 1, 2, and 3 → [1, 2, 3, 4].

In [60]:
a = [1,2,3,4,5,6]

a[:4]

[1, 2, 3, 4]

a is a list that contains different types of elements:

1, 2, 3 — integers

"string content" — a string

3+4j — a complex number (has a real part 3 and imaginary part 4)

Python lists can store mixed data types in the same list.

In [61]:
a = [1,2,3, "string content", 3+4j]

a

[1, 2, 3, 'string content', (3+4j)]

a is a list of strings, each representing a name or job title.

a[0] accesses the first element of the list because Python uses zero-based indexing (indexing starts at 0).

So a[0] returns "chandra".

In [62]:
a = ["chandra", "sai", "AI Engineer", "ML Engineer", "Data Science Engineer"]

a[0]

'chandra'

a is a list of strings.

a[0:2] is list slicing.

It gets elements from index 0 up to but not including index 2.

So it returns the first two items: "chandra" and "sai".

In [63]:
a = ["chandra", "sai", "AI Engineer", "ML Engineer", "Data Science Engineer"]

a[0:2]

['chandra', 'sai']

a is a list of strings.

a[:4] is a slice that means "start from the beginning (index 0) up to but not including index 4".

So it returns the first 4 items: "chandra", "sai", "AI Engineer", and "ML Engineer".

In [64]:
a = ["chandra", "sai", "AI Engineer", "ML Engineer", "Data Science Engineer"]

a[:4]

['chandra', 'sai', 'AI Engineer', 'ML Engineer']

a[0:4:2] is extended slicing with three parts: start:stop:step

start = 0: start from index 0 (the first element)

stop = 4: go up to but not include index 4

step = 2: take every 2nd element (skip 1 element in between)

So it takes elements at indices:

0 → "chandra"

2 → "AI Engineer"

Elements at indices 1 and 3 are skipped.

In [65]:
a = ["chandra", "sai", "AI Engineer", "ML Engineer", "Data Science Engineer"]

a[0:4:2]

# starting index=== 0
# last index======= 4
# skip 2 elements from it

['chandra', 'AI Engineer']

a[::2] is a slice with start and stop omitted, and a step of 2.

This means:

Start from the beginning (index 0),

Go all the way to the end,

Take every 2nd element (skip one in between).

So it picks elements at indices: 0, 2, 4.

In [66]:
a = ["chandra", "sai", "AI Engineer", "ML Engineer", "Data Science Engineer"]

a[::2]

['chandra', 'AI Engineer', 'Data Science Engineer']

a[-1] uses negative indexing.

Negative indices start counting from the end of the list:

-1 means the last element,

-2 means the second last, and so on.

So a[-1] returns "Data Science Engineer", the last item in the list.

In [67]:
a = ["chandra", "sai", "AI Engineer", "ML Engineer", "Data Science Engineer"]

a[-1]
# last element

'Data Science Engineer'

a[-2] uses negative indexing to access the second last element of the list.

Negative indices count from the end:

-1 is the last element ("Data Science Engineer")

-2 is the second last element ("ML Engineer")

So, a[-2] returns "ML Engineer".

In [68]:
a = ["chandra", "sai", "AI Engineer", "ML Engineer", "Data Science Engineer"]

a[-2]

'ML Engineer'

a[::-1] uses slice notation with a step of -1.

This means:

Start from the end of the list,

Move backwards by one element at a time,

Continue until the beginning of the list.

So, it reverses the list without modifying the original.

Output:
['Data Science Engineer', 'ML Engineer', 'AI Engineer', 'sai', 'chandra']

In [69]:
a = ["chandra", "sai", "AI Engineer", "ML Engineer", "Data Science Engineer"]

a[::-1]
# print in reverse order

['Data Science Engineer', 'ML Engineer', 'AI Engineer', 'sai', 'chandra']

a = [1] is a list with a single element 1.

a * 2 repeats the list twice.

So it creates a new list where the elements of a are repeated two times in a row.

Output:
[1, 1]


In [70]:
a = [1]

a * 2

# repeated elements twice

[1, 1]

a is a list containing three elements: [1, 2, 3].

a * 2 repeats the entire list twice.

It creates a new list where the sequence [1, 2, 3] appears twice, back to back.

Output:
[1, 2, 3, 1, 2, 3]

In [71]:
a = [1,2,3]

a * 2

# repeated elements twice

[1, 2, 3, 1, 2, 3]

a = [1, 2, 3]: This is a list with three elements.

a.append("new element"):

The .append() method adds a new item at the end of the list.

Here, "new element" is added as the last item.

After appending, a becomes [1, 2, 3, "new element"].

Output:
[1, 2, 3, 'new element']

In [72]:
a = [1,2,3]

a.append("new element")
# add element into list

a

[1, 2, 3, 'new element']

a = [1, 2, 3]: List with three numbers.

a.append([4, 5, 6]): The .append() method adds the entire list [4, 5, 6] as a single element to the end of a.

So after this, a becomes: [1, 2, 3, [4, 5, 6]].

Notice the last element is a nested list (a list inside a list).

In [73]:
a = [1,2,3]

a.append([4,5,6])

a

[1, 2, 3, [4, 5, 6]]

a = [1, 2, 3, 4, 5, 6]: A list with six elements.

a.pop(0):

The .pop() method removes and returns the element at the specified index.

Here, 0 means the first element (1) will be removed.

After popping, the list a becomes [2, 3, 4, 5, 6].

The removed element 1 is returned by .pop(0), but if you don't assign it to a variable or print it, it will just be discarded.

In [74]:
a = [1,2,3,4,5,6]

# pop ....zero index
a.pop(0)
a

[2, 3, 4, 5, 6]

Tuple

a = (1, 2, 3, 4, 5, 6) creates a tuple with 6 elements.

Tuples are similar to lists but are immutable (cannot be changed after creation).

type(a) returns the data type of a.

Since a is enclosed in parentheses (), Python treats it as a tuple.

In [75]:
a = (1,2,3,4,5,6)

type(a)

tuple

a is a tuple containing 4 string elements.

a[:2] is a slice that gets elements from the start (index 0) up to but not including index 2.

So it returns the first two elements: "first" and "second".

Slicing works the same way for tuples as it does for lists.
Output:
('first', 'second')

In [76]:
a = ("first", "second", "third","four")

a[:2]

('first', 'second')

a is a tuple with 4 elements.

a[-1] uses negative indexing to access the last element in the tuple.

Negative indices count from the end:

-1 is the last element ("four"),

-2 would be the second last, and so on.

So a[-1] returns "four".

Output:
'four'

In [77]:
a = ("first", "second", "third","four")

a[-1]

'four'

a is a tuple with 4 elements.

a[::-1] uses slice notation with a step of -1, meaning:

Start from the end,

Move backwards one element at a time,

Continue until the beginning.

This returns a new tuple with elements in reverse order.

Output:
('four', 'third', 'second', 'first')

In [78]:
a = ("first", "second", "third","four")

a[::-1]
# print in reverse order

('four', 'third', 'second', 'first')

t is a string containing mixed uppercase and lowercase letters.

t.lower() returns a new string where all characters are converted to lowercase.

The original string t remains unchanged because strings are immutable.

print(t.lower()) outputs the lowercase version of the string.

Output:
abxcdfbnght

In [79]:
t = "ABXCDFbnght"
t.lower()
print(t.lower())

#converted to LOWER CASE

abxcdfbnght


t is a string with all lowercase letters.

t.upper() converts all letters in the string to uppercase.

print(t.upper()) prints the uppercase version.

The original string t remains unchanged since strings are immutable.

Output:
ABCDEFGHIJK

In [80]:
t = "abcdefghijk"
print(t.upper())

# converted to UPPER CASE

ABCDEFGHIJK


t is a string "Hello World" with two words separated by a space.

t.split(' ') splits the string into a list of substrings using the space ' ' as the separator.

The result sliced is a list: ['Hello', 'World'].

This is useful to break a sentence or text into words.

Output:
['Hello', 'World']

In [81]:
t = "Hello World"
sliced = t.split(' ')

print(sliced)

# split data based on white-space

['Hello', 'World']


t is a string "Hello,World" where two words are separated by a comma.

t.split(',') splits the string into a list of substrings using the comma , as the separator.

The result sliced is a list: ['Hello', 'World'].

This lets you separate text based on commas or any other character you specify.

Output:
['Hello', 'World']

In [82]:
t = "Hello,World"
sliced = t.split(',')

print(sliced)

#split() ....the data based on comma

['Hello', 'World']


a = "first statement" initializes a string variable a.

a + " second-statement" concatenates (joins) the original string with " second-statement".

The result is assigned back to a, updating the variable to the combined string.

print(a) outputs the combined string.

Output:
first statement second-statement

In [83]:
a ="first statement"
a = a + " second-statement"

print(a)

# concatenate two strings

first statement second-statement


a is a tuple with 4 elements.

The slice a[0:3:2] means:

Start at index 0 (the first element: "first"),

Go up to (but not including) index 3 ("four" is at index 3, so stop before that),

Step by 2, which means take every second element.

So it takes elements at index 0 and 2: "first" and "third".

The element at index 1 ("second") is skipped because of the step 2.

Output:
('first', 'third')

In [84]:
a = ("first", "second", "third","four")

a[0:3:2]

# skip "second element" every time

('first', 'third')

a is a tuple with 10 string elements.

len(a) returns the total number of elements in the tuple, which is 10.

print("total elements length", len(a)) outputs:

total elements length 10


The slice a[0:9:2] means:

Start from index 0 (element "first"),

Go up to (but not including) index 9 (element "nine"),

Step by 2, which means pick every second element.

This skips every other element starting from the first one.

Output of the slice:
('first', 'third', 'five', 'seven', 'nine')

In [85]:
a = ("first", "second", "third","four","five", "six", "seven", "eight", "nine", "ten")
print("total elements length",len(a))
a[0:9:2]
#skip second element every time

total elements length 10


('first', 'third', 'five', 'seven', 'nine')

Dictionary

d is a dictionary in Python.

A dictionary stores key-value pairs.

Here, 'key' is the key (a string), and [1, 2, 3, 4] is the value (a list).

So d holds one entry where 'key' maps to the list [1, 2, 3, 4].

Dictionaries are useful for associating unique keys with data values.

Output:
{'key': [1, 2, 3, 4]}

In [86]:
d = {'key': [1,2,3,4]}

d

{'key': [1, 2, 3, 4]}

d["key"] accesses the value associated with the key "key" in the dictionary d.

Since earlier you defined:

d = {'key': [1, 2, 3, 4]}


this will return the list [1, 2, 3, 4].

In [87]:
d["key"]

[1, 2, 3, 4]

In Python dictionaries, keys must be unique.

If you try to create a dictionary with duplicate keys, Python will keep only the last value assigned to that key.

Here, you defined 'key' twice:

First: 'key': [1, 2, 3, 4]

Then: 'key': [5, 6, 7, 8]

Python replaces the first 'key' value with the second one.

So, the dictionary d actually becomes:

{'key': [5, 6, 7, 8]}

Output:
{'key': [5, 6, 7, 8]}

In [88]:
#get latest 'key' in "dictionary"
d = {'key': [1,2,3,4],
     'key': [5,6,7,8]}

d
# key has 1,2,3,4 ....these are replaced
 # python avoid duplicates in dictionary and replaces it last modified 'key'

{'key': [5, 6, 7, 8]}

d is a dictionary with two keys: 'key1' and 'key2'.

d.items (without parentheses) is just a reference to the method itself, not executed yet.

d.items() (with parentheses) calls the method and returns a view object containing all the dictionary’s key-value pairs as tuples.

Each item is a tuple like: ('key1', ["string1", "string2"]).

What happens:

d.items → refers to the method object (no output useful for data)

d.items() → returns something like:

dict_items([('key1', ['string1', 'string2']), ('key2', [1, 2, 3, 4, 5, 6])])

In [89]:
# print items in "dictionary"
d = {'key1': ["string1", "string2"],
     'key2': [1,2,3,4,5,6]
}

d.items
d.items()

dict_items([('key1', ['string1', 'string2']), ('key2', [1, 2, 3, 4, 5, 6])])

d.keys() is a method that returns a view object containing all the keys in the dictionary d.

It doesn’t return a list, but behaves like a set of keys that you can iterate over.

You can convert it to a list if you want to print keys as a list.

Example Output:
dict_keys(['key1', 'key2'])

In [90]:
#print keys in "dictionary"
d = {'key1': ["string1", "string2"],
     'key2': [1,2,3,4,5,6]
}

d.keys()
# print keys

dict_keys(['key1', 'key2'])

d.values() is a method that returns a view object containing all the values in the dictionary d.

Values are what each key in the dictionary points to.

It doesn’t return a list but behaves like a collection of the dictionary’s values.

You can convert it to a list if you want to see all values at once.

Example Output:
dict_values([['string1', 'string2'], [1, 2, 3, 4, 5, 6]])

In [91]:
d = {'key1': ["string1", "string2"],
     'key2': [1,2,3,4,5,6]
}

d.values()
#it returns all the values

dict_values([['string1', 'string2'], [1, 2, 3, 4, 5, 6]])

del d['key2'] deletes the key 'key2' and its associated value from the dictionary d.

After this operation, the dictionary d only contains 'key1' and its value.

del is a Python keyword used to remove variables, dictionary keys, list elements, etc.

Output:
{'key1': ['string1', 'string2']}

Summary:

Use del dict[key] to remove a key-value pair from a dictionary.

Trying to delete a key that doesn’t exist will raise a KeyError.

To avoid errors, you can check if the key exists before deleting.

In [92]:
# delete dictionry
d = {'key1': ["string1", "string2"],
     'key2': [1,2,3,4,5,6]
}

del d['key2']
# delete key2

d

{'key1': ['string1', 'string2']}

monuments is a dictionary inside a dictionary — also called a nested dictionary.

The outer dictionary has one key: "Taj Mahal".

The value for "Taj Mahal" is another dictionary: {"Location": "Agra", "Year": 1632}.

To access "Year" inside this nested dictionary, you use two keys:

First, get the value for "Taj Mahal": monuments["Taj Mahal"] → {"Location": "Agra", "Year": 1632}

Then, get "Year" from that dictionary: ["Year"]

So, monuments["Taj Mahal"]["Year"] gives you 1632.

Output:
1632

In [93]:
# Dictionary within Dictionary

monuments={"Taj Mahal":{"Location":"Agra","Year":1632}}

monuments["Taj Mahal"]["Year"]

1632

lang is a dictionary with states as keys and their languages as values.

The .get() method is used to retrieve the value for a given key.

Syntax: dictionary.get(key, default_value)

Here, "Goa" is not a key in the dictionary.

So, instead of causing an error like lang["Goa"] would, .get() returns the default value "Not Available".

This helps avoid errors when the key might be missing.

Output:
'Not Available'

Summary:

.get() safely fetches the value for a key.

If key is missing, returns the specified default instead of raising an error.

Very useful when working with uncertain dictionary data.

In [94]:
# get() function in "dictionary"
lang={"karnataka":"Kannada","Andhra Pradesh":"Telugu"}

lang.get("Goa","Not Available")

'Not Available'

The .pop() method removes a key from the dictionary and returns its value.

Here, "UK" is removed from capitals.

After pop, "UK": "London" is deleted from the dictionary.

It’s a safe way to remove a key and use the removed value if needed.

If the key doesn’t exist, .pop() will raise a KeyError unless a default value is provided as the second argument.

Output:
uk information got deleted
{'India': 'Delhi', 'Andhra Pradesh': 'Amaravati'}

Summary:

.pop(key) removes and returns the value for key.

Use it to delete entries while optionally using the removed value.

To avoid errors, you can provide a default value like .pop(key, default).

In [95]:
#pop() function in "dictionary"

capitals={"India":"Delhi","Andhra Pradesh":"Amaravati","UK":"London"}
capitals.pop("UK")
print("uk information got deleted")
print(capitals)

uk information got deleted
{'India': 'Delhi', 'Andhra Pradesh': 'Amaravati'}


The .popitem() method removes and returns the last inserted key-value pair from the dictionary.

Since Python 3.7+, dictionaries remember the order of insertion.

So, .popitem() removes the most recently added item.

It returns the removed item as a tuple (key, value).

If the dictionary is empty, calling .popitem() raises a KeyError.

Example:
capitals = {"India": "Delhi", "Andhra Pradesh": "Amaravati", "UK": "London"}

removed_item = capitals.popitem()
print("Removed:", removed_item)
print("Remaining dict:", capitals)


Possible Output:

Removed: ('UK', 'London')
Remaining dict: {'India': 'Delhi', 'Andhra Pradesh': 'Amaravati'}

Summary:

Use .popitem() to remove the last added key-value pair from a dictionary.

Useful for stack-like behavior or when order matters.

Raises error if dictionary is empty.

In [96]:
# delete the popitem() in dictionary

capitals.popitem()

('Andhra Pradesh', 'Amaravati')

The .clear() method removes all items from the dictionary, making it empty.

After calling capitals.clear(), the dictionary capitals becomes {}.

The .clear() method does not return anything (returns None), so printing capitals.clear() will print None.

You only need to call .clear() once to empty the dictionary.

Example:
capitals = {"India": "Delhi", "Andhra Pradesh": "Amaravati", "UK": "London"}

capitals.clear()
print(capitals)         # Prints: {}
print(capitals.clear()) # Prints: None

Output:
{}
None
------
Summary:

.clear() empties the dictionary.

It changes the dictionary in place.

It returns None — don’t expect any output when you print the result of .clear().

In [97]:
# clear() function in dictionary

capitals.clear()
print(capitals.clear())

None


states and capital are two separate lists.

zip(states, capital) pairs elements from both lists based on their positions.

Example: ("Rajasthan", "Jaipur"), ("Assam", "Dispur"), etc.

list(zip(states, capital)) converts these pairs into a list of tuples.

dict(zip(states, capital)) converts the zipped pairs into a dictionary, where:

Each element from states becomes a key.

Each corresponding element from capital becomes the value for that key.

This is a common way to create a dictionary when you have two related lists.

Output:
sample list:  [('Rajasthan', 'Jaipur'), ('Assam', 'Dispur'), ('Bihar', 'Patna')]
sample dictionary {'Rajasthan': 'Jaipur', 'Assam': 'Dispur', 'Bihar': 'Patna'}

Summary:

zip() pairs items from multiple lists.

dict(zip()) quickly creates dictionaries from two lists.

Useful for mapping related data like states and their capitals.

In [98]:
#convert two lists into Dictionary

states=["Rajasthan","Assam","Bihar"]
capital=["Jaipur","Dispur","Patna"]

zip(states,capital)
list(zip(states,capital))
print("sample list: ",list(zip(states,capital)))
print("sample dictionary",dict(zip(states,capital)))

sample list:  [('Rajasthan', 'Jaipur'), ('Assam', 'Dispur'), ('Bihar', 'Patna')]
sample dictionary {'Rajasthan': 'Jaipur', 'Assam': 'Dispur', 'Bihar': 'Patna'}


This defines a function named factorial that calculates the factorial of a number using recursion.

Factorial of a number n (written as n!) is the product of all positive integers from 1 to n.

Special case: 0! is defined as 1.

How the function works:

Base Case:

When number == 0, it returns 1.

This stops the recursion.

Recursive Case:

For any other number, it returns number * factorial(number - 1).

This means the function calls itself with a smaller number, multiplying the current number by the factorial of one less than it.

In [99]:
def factorial(number):
    if number==0:
        return 1
    else:
        return number * factorial(number-1) 

factorial(0)        

# function

1

In [100]:
factorial(5)

120

Explanation:

You're passing a string ("3") instead of an integer to the factorial function.

Problem:

The function expects a number (int), but "3" is a string.

When Python tries to evaluate "3" == 0 or "3" - 1, it will raise a TypeError.

In [101]:
factorial("3")

TypeError: unsupported operand type(s) for -: 'str' and 'int'

In Python, you can write a single-line comment using the # symbol. Anything written after # on that line will be ignored by the Python interpreter.

In [ ]:
#single comment line

✅ 1. # function definition

This is a comment.

The # symbol tells Python to ignore this line.

It’s used here to indicate that the next block defines a function.

✅ 2. def add(num1, num2):

def is the keyword used to define a function.

add is the name of the function. You can call it anything, but names should be meaningful.

(num1, num2) are the parameters of the function — they accept values when the function is called.

: marks the beginning of the function's body (the indented block that follows).

✅ 3. return num1 + num2

This is the body of the function.

The return statement sends the result of num1 + num2 back to the caller.

In this case, it adds the two numbers passed in and returns the sum.

For example, if num1 = 5 and num2 = 5, the function returns 10.

✅ 4. # function calling

Another comment, used here to label the part of the code where the function is being called.

✅ 5. add(5, 5)

This is a function call.

You're passing two arguments: 5 and 5 to the add() function.

Internally, num1 = 5 and num2 = 5.

It returns 10, but since you’re not using print() or assigning it to a variable, the result is computed and discarded.

🔁 To See the Output

If you want to see the result, you should either:

Print it:

print(add(5, 5))  # Output: 10


Or store it in a variable:

result = add(5, 5)
print(result)     # Output: 10


In [ ]:
# function definition
def add(num1,num2):
    return num1 + num2

#function calling
add(5,5)

10

📌 def function():

def: Keyword used to define a function.

function: The name of your function (not a good name in practice — more on that later).

():: Parentheses indicate that the function takes no parameters.

:: Starts the indented block (the body of the function).

📌 """ doc string ... """

This is a docstring — short for documentation string.

✅ What is a docstring?

It's a string literal that appears as the first statement in a function, class, or module.

Python uses this string as official documentation.

In [ ]:
def function():
    """ 
     doc string 
    
    to describe the purpose of function
    """
    pass 

function()

In [ ]:
# interpreter encounters "pass" keyword, it does nothing and moves on to the next line. 

In [ ]:
#function definition 
def add(number1, number2):
    """ it takes two arguments
    first parameter: number1
    second parameter: number2

    Returns: finally adds the two number and returns the output
    """
    return number1 + number2 # function definition ends here

#function calling....to execute the function
add(20, 5)    

25

*args collects all positional arguments passed to the function into a tuple.

You then convert args to a list (which is not necessary in this case) and iterate through it.

Each argument is printed one by one.

In [ ]:
#args .....within function()
def argument_function(*args):
    for i in list(args):
        print(i)


argument_function(2,3,4,5)

2
3
4
5


*args allows the function to accept a variable number of arguments.

sum is initialized to 0.

The function iterates over all the values in args and adds them.

Finally, it prints the result.

💡 Output:
sum of all values 21

In [ ]:
def add_arguments(*args):
    sum=0
    for i in args:
        sum+=i
    print("sum of all values",sum)    

add_arguments(1,2,3,4,5,6)    

sum of all values 21


💡 Example Use-Cases for **kwargs:

Flexible APIs

Logging or debugging info

Passing config settings dynamically

In [ ]:
#kwargs.....
def function(**kwargs):
    return kwargs

function(name="chandra",email_id="chandrasekharcse522@gmail.com")

{'name': 'chandra', 'email_id': 'chandrasekharcse522@gmail.com'}

Lambda Functions in Python
What is a lambda function?

A lambda function is a small anonymous function (a function without a name).

It can take any number of arguments but only has one expression.

The expression is evaluated and returned automatically.

Syntax of lambda function
lambda arguments: expression


lambda is the keyword.

arguments is a comma-separated list of parameters.

expression is a single expression (no statements allowed).
Output: The result 6 is returned.

In [ ]:
#lambda

func= lambda x: x+1

#pass value to func()
func(5)

6

lambda a, b:
This lambda function takes two parameters: a and b.

a + b
The expression returns the sum of a and b.

func2(5, 5)
Calling the function with arguments 5 and 5, so it returns 5 + 5 = 10.

print(result)
Prints the output, which is 10.

In [ ]:
#lambda
func2 = lambda a,b : a+b

func2(5,5)

10

lambda a:
This defines a lambda function with one parameter a.

a ** 2
This is the expression that calculates the square of a (i.e., a raised to the power 2).

func3(5)
Calls the lambda function with argument 5.

The function computes 
 and returns 25.

print(result) outputs the result, which is 25.

In [ ]:
#lambda
#square of the function()
func3= lambda a: a**2

func3(5)

25

In [ ]:
#lambda
#max value
func4 = lambda a,b: a if a>b else b

func4(5,27)

27

In [ ]:
#lambda
#min value
func5= lambda a,b: b if a>b else a


func5(27,3)

3

In [ ]:
# how the code gets executed?

print("number1")
print("second number")
for i in range(3):
    print(i)
print("third number")    

number1
second number
0
1
2
third number


In [ ]:
df1 = 5
type(df1)

int

In [ ]:
# type conversion
# convert integer to string

vari = str(df1)
type(vari)

str

In [ ]:
b1= input("enter number")
print(b1)

6


In [ ]:
a1 = 20
a1 + b1

# type(b)====str
#trying to add int and str

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [ ]:
xyz= input("enter user input")
print("given input",xyz)

type(xyz)

print("data type :", type(xyz))

given input string
data type : <class 'str'>


In [ ]:
a=20
#input()....... asking user to "enter a number"
b= int(input("Ener number")) # converted it into integer
print("b data type is",type(b))
print(a + b)


b data type is <class 'int'>
26


In [ ]:
v= "chandra"
print(type(v))
print(v)
list(v)
print("converted to list data type",list(v))


<class 'str'>
chandra
converted to list data type ['c', 'h', 'a', 'n', 'd', 'r', 'a']


In [ ]:
name=" chandra"
title="AI Engineer"
print(f"my name is: {name} and my profession is: {title}")

my name is:  chandra and my profession is: AI Engineer


In [ ]:
expression = input("Enter expression")
result = eval(expression)
print(result)

# Enter expression   3 + 7
#given input is string data type
# eval ()


10


In [ ]:
type(result)

int

In [ ]:
a = eval("2 * 5")
a

10

In [ ]:
a = eval("3 + 5")
a

8

In [ ]:
a = eval("2 + 3 + 5")
a

10

In [ ]:
a = eval("2 * 4 * 10")
a

80

In [ ]:
a = eval("100 / 5")
a

20.0

In [ ]:
len("hello world")

11

In [ ]:
b= eval("len('hello world')")
print(b)
type(b)

11


int

for loop

In [ ]:
a = [1,2,3,4,5,6]

for i in a:
    print(i)

1
2
3
4
5
6


In [ ]:
t = (1,2,3,4)

for i in t:
    print(i)

1
2
3
4


In [ ]:
fruits = ['apple', 'banana', 'grapes']

for index, i in enumerate(fruits):
    print(index,i)

0 apple
1 banana
2 grapes


In [ ]:
# function to print even numbers

def even_number_sum():
    sum=0
    numbers = [1,2,3,4,5,6,7,8,9,10]
    for i in numbers:
        if i%2==0:
         sum+=i
    print(sum) 

even_number_sum()       

30


In [ ]:
#last occurence of target character
text = "iT is a text CONTENT and prints a value"
target_value='a'
counter = 0
print("length:",len(text))
for i in range(len(text)):
    if text[i].lower()==target_value:
        counter=i
        


print(counter)




length: 39
35


In [ ]:
#sum of all numbers within range of 50
sum=0
for i in range(1,50):
    sum=sum + i
print(f"the total sum is:{sum}")

the total sum is:1225


In [ ]:
#sum of all even numbers within range of 50
sum=0
for i in range(1,50):
    if i%2==0:  #even number condition
        sum+=i
print(f"sum of all even numbers within range of 50:{sum}")

sum of all even numbers within range of 50:600


In [ ]:
#sum of all odd numbers within range of 50
sum=0
for i in range(1,50):
    if i%2!=0:  #odd number condition
        sum+=i
print(f"sum of all odd numbers within range of 50:{sum}")

sum of all odd numbers within range of 50:625


In [ ]:
# first index of even integer within range of 50
position=0
for i in range(1,50):
    if i%2==0:
        position=position+1
        break
print(f"the first index of even integer is {position}")    

the first index of even integer is 1


In [ ]:
# last index of  even integer within range of 50
position=0
for i in range(1,50):
    if i%2==0:
        position=position+1
        continue
print(f"the last index of even integer is {position}") 

the last index of even integer is 24


In [ ]:
# first index of odd integer within range of 50
position=0
for i in range(1,50):
    if i%2!=0:
        position=position+1
        break
print(f"the first index of odd integer is {position}") 

the first index of odd integer is 1


In [ ]:
# last index of odd integer within range of 50
position=0
for i in range(1,50):
    if i%2!=0:
        position=position+1
        continue
print(f"the last index of odd integer is {position}") 

the last index of odd integer is 25


In [ ]:
# print the first even integer within range of 50
first_even_integer = 0
for i in range(1,50):
    if i%2==0:
        first_even_integer= i
        break
print(first_even_integer)

2


In [ ]:
# print the first odd integer within range of 50
first_odd_integer = 0
for i in range(1,50):
    if i%2!=0:
        first_odd_integer= i
        break
print(first_odd_integer)

1


In [ ]:
# print the last even integer within range of 50
last_even_integer = 0
for i in range(1,50):
    if i%2==0:
        last_even_integer= i
        continue
print(last_even_integer)

48


In [ ]:
# print the last odd integer within range of 50
last_odd_integer = 0
for i in range(1,50):
    if i%2!=0:
        last_odd_integer= i
        continue
print(last_odd_integer)

49


In [ ]:
# print prime numbers
for num in range(5,15):
    if num > 1:
        for i in range(2, num):
            if num%i == 0:
                break
        else:
            print(num)

5
7
11
13


In [ ]:
#reverse a number
number= 678901
reverse=0
while number > 0:
    remainder = number%10
    reverse= remainder+(reverse * 10)
    number = number//10

print(reverse)    

109876


In [ ]:
# print all the digits within a number
sum=0
number= 123456
while number>0:
    remainder=number%10
    sum+= remainder
    number//=10
print(sum)    

21


In [ ]:
#print datetime
import datetime

print(datetime.datetime.now())

2025-09-17 12:07:20.324212


In [ ]:
#print date
import datetime

print(datetime.date.today())

2025-09-17


In [ ]:
#print day
import datetime

print(datetime.date.today().day )

17


In [ ]:
# add 7 days to the current date
import datetime

 
print(datetime.date.today() + datetime.timedelta(days=7))

2025-09-24


In [ ]:
#print last month
import datetime

last_month = datetime.date.today() - datetime.timedelta(days=30)
print(last_month)

2025-08-18


In [ ]:
#print hour
import datetime

print(datetime.datetime.now().hour)

12


In [ ]:
#print minute
import datetime

print(datetime.datetime.now().minute)

14


In [ ]:
#print second
import datetime

print(datetime.datetime.now().second)

36


In [ ]:
import datetime

# Get the current date and time
now = datetime.datetime.now()

print("--- Current Date and Time ---")
print("Full datetime object:", now)
print("Year:", now.year)
print("Month:", now.month)
print("Day:", now.day)
print("Hour:", now.hour)
print("Minute:", now.minute)
print("Second:", now.second)
print("Microsecond:", now.microsecond)

--- Current Date and Time ---
Full datetime object: 2025-09-17 12:24:47.596869
Year: 2025
Month: 9
Day: 17
Hour: 12
Minute: 24
Second: 47
Microsecond: 596869


In [ ]:
import datetime

now = datetime.datetime.now()

print("--- Formatted Date and Time ---")
print("Long format:", now.strftime("%A, %B %d, %Y"))
print("Short format:", now.strftime("%Y-%m-%d %H:%M:%S"))
print("Month and Year:", now.strftime("%B %Y"))
print("Just the time:", now.strftime("%I:%M %p")) # 12-hour clock with AM/PM

--- Formatted Date and Time ---
Long format: Wednesday, September 17, 2025
Short format: 2025-09-17 12:26:21
Month and Year: September 2025
Just the time: 12:26 PM


In [ ]:
import time

print("Starting...")
time.sleep(3)  # wait for 3 seconds
print("Done waiting!")


Starting...
Done waiting!


In [ ]:
import time

start = time.time()
# some task
for _ in range(100):
    pass
end = time.time()

print("Time taken:", end - start, "seconds")

#used in checking how much time machine learning model is taken to execute



Time taken: 0.0 seconds


In [ ]:
#CPU time.........time.perf_counter() gives you a high-precision timer — great for measuring the actual runtime of code with sub-second accuracy (better than time.time() for this).
import time
import random

def slow_task():
    # Create a big list of random numbers and sort it
    numbers = [random.randint(1, 1000000) for _ in range(1000000)]
    numbers.sort()

# Start timer
start = time.perf_counter()

# Run the task
slow_task()

# End timer
end = time.perf_counter()

# Calculate elapsed time
elapsed = end - start
print(f"Task completed in {elapsed:.4f} seconds")


Task completed in 4.1274 seconds


In [ ]:
#zip() function and enumerate() function
list1=[1,2,3]
list2=['chandra','sai','ML Engineer']
list3=['Data Science','MLOPS']
zipped_content=zip(list1,list2,list3)

print(zipped_content)
print(f"print zipped content in the form of list: {list(zipped_content)}")
print("")
print(f"dictionaryy format of zipped content{dict(zipped_content)}")

print zipped content in the form of list: [(1, 'chandra', 'Data Science'), (2, 'sai', 'MLOPS')]

dictionaryy format of zipped content{}


In [ ]:
#zipped content for dictionary type cast of data type ....empty{}
#change the above code and try to print dict type data
#zip() function and enumerate() function
list1=[1,2,3]
list2=['chandra','sai','ML Engineer']
list3=['Data Science','MLOPS']
zipped_content=zip(list1,list2,list3)

print(zipped_content)
print(f"print zipped content in the form of list: {list(zipped_content)}")
print("")
list11=[1,2,3]
list22=['chandra','sai','ML Engineer']
#list33=['Data Science','MLOPS','']
zipped_content1=zip(list11,list22)
print(f"dictionaryy format of zipped content{dict(zipped_content1)}")

print zipped content in the form of list: [(1, 'chandra', 'Data Science'), (2, 'sai', 'MLOPS')]

dictionaryy format of zipped content{1: 'chandra', 2: 'sai', 3: 'ML Engineer'}


In [ ]:
#set
l=[1,2,3,4]
list5=set(l)
list5

{1, 2, 3, 4}

In [ ]:
#set
example=['chandra','list content']
set_content=set(example)
set_content

{'chandra', 'list content'}

In [ ]:
#set example it stores unique values 
example=('chandra', 'chandra','chandra')
set_example=set(example)
list_ex=[1,1,1,1,1,1,1,1,1,1,1,1,]
set_li=set(list_ex)
set_example,set_li

({'chandra'}, {1})

In [ ]:
#set example
s1={10,9,8,7,6,5,3,2,1}

#display the data in sorted ascending order
s1

{1, 2, 3, 5, 6, 7, 8, 9, 10}

In [ ]:
#add value to the set
s12={10,9,7,3,2,1}
s12.add(343)

s12

{1, 2, 3, 7, 9, 10, 343}

In [ ]:
class Myclass:
    
    def basic(self):
        print("my class")

c=Myclass()

c.basic()

my class


In [ ]:
#closure 

def outer():
    x=10
    def inner():
        print(x)

    return inner

closure=outer()

#function calling
closure()

10


In [ ]:
#closure

def create_counter():
    count = 0
    def increment():
        nonlocal count  # The nonlocal keyword is needed to modify the variable
        count += 1
        return count
    return increment

counter1 = create_counter()
print(f"Counter 1: {counter1()}") # 1
print(f"Counter 1: {counter1()}") # 2

counter2 = create_counter()
print(f"Counter 2: {counter2()}") # 1

Counter 1: 1
Counter 1: 2
Counter 2: 1


In [ ]:
#decorators
# a decorator is a closure that takes a function as an argument and returns a new function.

def my_decorator(function):
    def wrapper():
        print("before the function")
        function()
        print("after function ")
    return wrapper


@my_decorator
def weather_api():
    print("Weather information")


#function calling
weather_api()



before the function
Weather information
after function 


In [ ]:
# List comprehension
# copy the numbers from range of 0 to 10 within new list


number_2=[number for number in range(0,11)]
print(number_2)

## actually what this list comprehension does
#number_2=[]
# for i in range(0,11)
#     number_2.append(i)
#print(number_2)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [ ]:
# use "List Comprehension"
# copy even numbers from the range of 1 to 10

numbers=[number for number in range(1,11) if number%2==0]
print(numbers)

[2, 4, 6, 8, 10]


In [ ]:
# use "List Comprehension"
#copy the odd numbers from the range of 1 to 10

numbers=[number for number in range(1,11) if number % 2!=0]
print(numbers)

[1, 3, 5, 7, 9]


In [ ]:
# use "List Comprehension"
# print the square of the numbers from range 1 to 10
numbers=[number**2 for number in range(1,11)]
print(numbers)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [ ]:
# List Comprehension
# using if and else block
# print even or odd from the range of 1 to 10

numbers=["even" if number%2==0 else "odd" for number in range(1,11) ]
print(numbers)

['odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even']


In [ ]:
# List Comprehension
# using function() within list comprehension

def square(number):
    return number**2

numbers=[square(number) for number in range(1,11)]
print(numbers)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [ ]:
# string operations within List Comprehension

content=["hello","chandra"]

output=[sample.upper() for sample in content]
print(output)

['HELLO', 'CHANDRA']


In [ ]:
# zip() function using "List Comprehension"

names=["Chandra","Sai"]
scores=[85, 90]

combined=[f"{n}: {s}" for n,s in zip(names,scores)]
print(combined)

['Chandra: 85', 'Sai: 90']


In [ ]:
# enumerate() function within "List Comprehension"

data=["a","b","c"]
indexed=[(i,val) for i,val in enumerate(data)]
print(indexed)

[(0, 'a'), (1, 'b'), (2, 'c')]


In [ ]:
# try except

try:
    1/0
except:
    print("Some Error")    

Some Error


In [ ]:
# print the exception
try:
    1/0
except Exception as e:
    print(e)    

division by zero


In [ ]:
try:
    1/0
except ZeroDivisionError as e:
    print(e)    

division by zero


In [ ]:
try:
    num=int(input("Enter number"))
except Exception as e:
    print(e)

invalid literal for int() with base 10: '"chandra"'


In [ ]:
try:
    num=int(input("Enter number"))
except ValueError as e:
    print(e)    

invalid literal for int() with base 10: '"chandra"'


In [ ]:
# try and except and finaly

try:
    num = int(input("Enter a number: "))
    print(f"You entered: {num}")
except ValueError:
    print("Oops! That's not a valid number.")
finally:
    print("This always runs, no matter what.")


Oops! That's not a valid number.
This always runs, no matter what.
